File: FluteModesAndroid/app/src/main/java/com/seagomezar/flutemodes/model/MusicalMode.kt
package com.seagomezar.flutemodes.model

enum class ModeFamily(val displayName: String) {
    MAJOR("Familia Mayor"),
    MINOR("Familia Menor")
}

enum class ModeType(
    val id: Int,
    val family: ModeFamily,
    val intervals: List<Int>, // semitone intervals from root
    val formula: String,
    val character: String
) {
    IONIAN(1, ModeFamily.MAJOR, listOf(0, 2, 4, 5, 7, 9, 11), "1 - 2 - 3 - 4 - 5 - 6 - 7", "Brillante y luminoso"),
    LYDIAN(2, ModeFamily.MAJOR, listOf(0, 2, 4, 6, 7, 9, 11), "1 - 2 - 3 - ♯4 - 5 - 6 - 7", "Etéreo y soñador"),
    MIXOLYDIAN(3, ModeFamily.MAJOR, listOf(0, 2, 4, 5, 7, 9, 10), "1 - 2 - 3 - 4 - 5 - 6 - ♭7", "Festivo y folclórico"),
    DORIAN(4, ModeFamily.MINOR, listOf(0, 2, 3, 5, 7, 9, 10), "1 - 2 - ♭3 - 4 - 5 - 6 - ♭7", "Melancólico y noble"),
    AEOLIAN(5, ModeFamily.MINOR, listOf(0, 2, 3, 5, 7, 8, 10), "1 - 2 - ♭3 - 4 - 5 - ♭6 - ♭7", "Triste e introspectivo"),
    PHRYGIAN(6, ModeFamily.MINOR, listOf(0, 1, 3, 5, 7, 8, 10), "1 - ♭2 - ♭3 - 4 - 5 - ♭6 - ♭7", "Oscuro y exótico"),
    LOCRIAN(7, ModeFamily.MINOR, listOf(0, 1, 3, 5, 6, 8, 10), "1 - ♭2 - ♭3 - 4 - ♭5 - ♭6 - ♭7", "Inestable y tenso");

    fun getName(loc: LocalizationManager): String {
        return when (this) {
            IONIAN -> loc.t("mode_1_name")
            LYDIAN -> loc.t("mode_2_name")
            MIXOLYDIAN -> loc.t("mode_3_name")
            DORIAN -> loc.t("mode_4_name")
            AEOLIAN -> loc.t("mode_5_name")
            PHRYGIAN -> loc.t("mode_6_name")
            LOCRIAN -> loc.t("mode_7_name")
        }
    }

    fun accidentalBadge(loc: LocalizationManager): String {
        return when (this) {
            IONIAN -> loc.t("acc_ionian")
            LYDIAN -> loc.t("acc_lydian")
            MIXOLYDIAN -> loc.t("acc_mixolydian")
            DORIAN -> loc.t("acc_dorian")
            AEOLIAN -> loc.t("acc_aeolian")
            PHRYGIAN -> loc.t("acc_phrygian")
            LOCRIAN -> loc.t("acc_locrian")
        }
    }

    fun keySignature(tonic: Tonic): String {
        return when (family) {
            ModeFamily.MAJOR -> tonic.majorKey
            ModeFamily.MINOR -> tonic.minorKey
        }
    }

    fun keyDescription(tonic: Tonic): String {
        return when (family) {
            ModeFamily.MAJOR -> "Armadura fija: ${tonic.majorKey} Mayor (${tonic.majorDesc})"
            ModeFamily.MINOR -> "Armadura fija: ${tonic.minorKey} Menor (${tonic.minorDesc})"
        }
    }
}

File: FluteModes/FluteModes/Models/MusicalMode.swift
import Foundation

public enum ModeFamily: String, CaseIterable, Identifiable, Codable {
    case major = "Familia Mayor"
    case minor = "Familia Menor"

    public var id: String { rawValue }
}

public enum ModeType: Int, CaseIterable, Identifiable, Codable {
    case ionian = 1
    case lydian = 2
    case mixolydian = 3
    case dorian = 4
    case aeolian = 5
    case phrygian = 6
    case locrian = 7

    public var id: Int { rawValue }

    public var family: ModeFamily {
        switch self {
        case .ionian, .lydian, .mixolydian:
            return .major
        case .dorian, .aeolian, .phrygian, .locrian:
            return .minor
        }
    }

    public var intervals: [Int] {
        switch self {
        case .ionian:     return [0, 2, 4, 5, 7, 9, 11]
        case .lydian:     return [0, 2, 4, 6, 7, 9, 11]
        case .mixolydian: return [0, 2, 4, 5, 7, 9, 10]
        case .dorian:     return [0, 2, 3, 5, 7, 9, 10]
        case .aeolian:    return [0, 2, 3, 5, 7, 8, 10]
        case .phrygian:   return [0, 1, 3, 5, 7, 8, 10]
        case .locrian:    return [0, 1, 3, 5, 6, 8, 10]
        }
    }

    public var name: String {
        switch self {
        case .ionian: return LocalizationManager.shared.t("mode_1_name")
        case .lydian: return LocalizationManager.shared.t("mode_2_name")
        case .mixolydian: return LocalizationManager.shared.t("mode_3_name")
        case .dorian: return LocalizationManager.shared.t("mode_4_name")
        case .aeolian: return LocalizationManager.shared.t("mode_5_name")
        case .phrygian: return LocalizationManager.shared.t("mode_6_name")
        case .locrian: return LocalizationManager.shared.t("mode_7_name")
        }
    }

    public var character: String {
        switch self {
        case .ionian: return "Brillante y luminoso"
        case .lydian: return "Etéreo y soñador"
        case .mixolydian: return "Festivo y folclórico"
        case .dorian: return "Melancólico y noble"
        case .aeolian: return "Triste e introspectivo"
        case .phrygian: return "Oscuro y exótico"
        case .locrian: return "Inestable y tenso"
        }
    }

    public var formula: String {
        switch self {
        case .ionian: return "1 - 2 - 3 - 4 - 5 - 6 - 7"
        case .lydian: return "1 - 2 - 3 - ♯4 - 5 - 6 - 7"
        case .mixolydian: return "1 - 2 - 3 - 4 - 5 - 6 - ♭7"
        case .dorian: return "1 - 2 - ♭3 - 4 - 5 - 6 - ♭7"
        case .aeolian: return "1 - 2 - ♭3 - 4 - 5 - ♭6 - ♭7"
        case .phrygian: return "1 - ♭2 - ♭3 - 4 - 5 - ♭6 - ♭7"
        case .locrian: return "1 - ♭2 - ♭3 - 4 - ♭5 - ♭6 - ♭7"
        }
    }

    public func accidentalBadge(for tonic: Tonic) -> String {
        switch self {
        case .ionian:
            return LocalizationManager.shared.t("acc_ionian")
        case .lydian:
            return LocalizationManager.shared.t("acc_lydian")
        case .mixolydian:
            return LocalizationManager.shared.t("acc_mixolydian")
        case .dorian:
            return LocalizationManager.shared.t("acc_dorian")
        case .aeolian:
            return LocalizationManager.shared.t("acc_aeolian")
        case .phrygian:
            return LocalizationManager.shared.t("acc_phrygian")
        case .locrian:
            return LocalizationManager.shared.t("acc_locrian")
        }
    }

    public func keySignature(for tonic: Tonic) -> String {
        switch family {
        case .major:
            return tonic.majorKey
        case .minor:
            return tonic.minorKey
        }
    }

    public func keyDescription(for tonic: Tonic) -> String {
        switch family {
        case .major:
            return "Armadura fija: \(tonic.majorKey) Mayor (\(tonic.majorDesc))"
        case .minor:
            return "Armadura fija: \(tonic.minorKey) Menor (\(tonic.minorDesc))"
        }
    }
}
